### ISSUE TITLE
Architecture of PPO LSTM

### ISSUE THREAD

**Issue by anilkurkcu** [Stable-Baselines-Team/stable-baselines3-contrib#183]
### ❓ Question

Hi, could you please explain how the default architecture looks like in the PPO LSTM implementation?

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/issues) in the repo
- [X] I have read the [documentation](https://sb3-contrib.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Hello,
what is your current understanding?
(i think there a similar issue already)

**Comment by anilkurkcu** [comment]
I understand it as similar to PPO implementation without LSTM, where 2 hidden layers of 64 dimension are used.

Could you share the issue maybe?

**Comment by araffin** [comment]
Related to https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/issues/160#issuecomment-1461150656  https://github.com/DLR-RM/stable-baselines3/issues/1005 and https://github.com/DLR-RM/stable-baselines3/issues/329


> I understand it as similar to PPO implementation without LSTM, where 2 hidden layers of 64 dimension are used.

Yes with an additional LSTM layers for each of the actor and the critic.

https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/21cc96cafd77a3a4347b43c374a4cd21bf08d804/sb3_contrib/common/recurrent/policies.py#L83-L86

```python
from sb3_contrib import RecurrentPPO

RecurrentPPO("MlpLstmPolicy", "CartPole-v1").policy
```

```
RecurrentActorCriticPolicy(
  (features_extractor): FlattenExtractor(
    (flatten): Flatten(start_dim=1, end_dim=-1)
  )
  (pi_features_extractor): FlattenExtractor(
    (flatten): Flatten(start_dim=1, end_dim=-1)
  )
  (vf_features_extractor): FlattenExtractor(
    (flatten): Flatten(start_dim=1, end_dim=-1)
  )
  (mlp_extractor): MlpExtractor(
    (policy_net): Sequential(
      (0): Linear(in_features=256, out_features=64, bias=True)
      (1): Tanh()
      (2): Linear(in_features=64, out_features=64, bias=True)
      (3): Tanh()
    )
    (value_net): Sequential(
      (0): Linear(in_features=256, out_features=64, bias=True)
      (1): Tanh()
      (2): Linear(in_features=64, out_features=64, bias=True)
      (3): Tanh()
    )
  )
  (action_net): Linear(in_features=64, out_features=2, bias=True)
  (value_net): Linear(in_features=64, out_features=1, bias=True)
  (lstm_actor): LSTM(4, 256)
  (lstm_critic): LSTM(4, 256)
)
```

The full logic:
https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/21cc96cafd77a3a4347b43c374a4cd21bf08d804/sb3_contrib/common/recurrent/policies.py#L230-L257

**Comment by fontanafederico** [comment]
Hi @araffin, follow-up question on the design choice of how sequences are handled with LSTM. 

Following the naming convention of [torch.nn.LSTM](https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html), let's say that we are dealing with an observation space Box of shape (20, 10), where the sequence length L is 20 and the input_size is 10. In other words, at each steps the agent is given 10 time series of equal length 20 each. Due to `FlattenExtractor`, `RecurrentPPO.policy.lstm_actor` will be LSTM(200) and therefore the sequence length is shrunk from L=20 to L=1, and input size went from 10 to 200. This seems to be confirmed by the source code here:

https://github.com/Stable-Baselines-Team/stable-baselines3-contrib/blob/35f06254ba3a470e8bbe1dff0bfc6f319ee2431a/sb3_contrib/common/recurrent/policies.py#L179C1-L187C1

I'm happy to be proven wrong, but flattening the input will discard relevant information from the input and increase the number of parameters of LSTM. Not flattening may (or may not) lead to simpler model and faster learning thanks to preserving the sequential nature of the input. Put in other words, `RecurrentActorCriticPolicy` does not currently distinguish from on observation space of size (20, 10) vs (1, 200). While this certainly make sense for MLP, what's the rationale for why it holds true for LSTM policies?

**Comment by araffin** [comment]
>  let's say that we are dealing with an observation space Box of shape (20, 10), where the sequence length L is 20 and the input_size is 10

Hello,
I think there is a misconception of what the PPO LSTM does in SB3.
The sequence considered is a sequence of transitions (obs, action, reward, next obs), not a sequence from the input obs (so in your case, PPO LSTM expect a Box(10,), the sequence is created automatically and only depends on episodes.
